### PR TITLE
Add context argument to IndexReader.Postings

### PR DESCRIPTION
**Pull request by aknuds1**
Add context argument to `tsdb.IndexReader.Postings`. This is in preparation for `tsdb.PostingsForMatchers` [gaining a context argument](https://github.com/prometheus/prometheus/pull/12359), the motivation being to respect context cancellation.

Part of https://github.com/prometheus/prometheus/issues/12836.

### CODE REVIEW

**Comment by aknuds1** [comment]
I think the failing test is a flake (that test tends to flake on PRs).

**Review by jesusvazquez — APPROVED** [review]
Thank you for this contribution @aknuds1

File: tsdb/index/postings.go [hunk 1]
```diff
@@ -559,12 +560,14 @@ type mergedPostings struct {
...
 
-func newMergedPostings(p []Postings) (m *mergedPostings, nonEmpty bool) {
+func newMergedPostings(ctx context.Context, p []Postings) (m *mergedPostings, nonEmpty bool) {
 	ph := make(postingsHeap, 0, len(p))
 
 	for _, it := range p {
 		// NOTE: mergedPostings struct requires the user to issue an initial Next.
 		switch {
+		case ctx.Err() != nil:
+			return &mergedPostings{err: ctx.Err()}, true
```
Comment (bboreham) [inline tsdb/index/postings.go:570]: I don't think this check is really necessary - this loop is doing an initial `Next()` on each of the postings, but the main work is done on each subsequent call to `Next()` on the merged structure.

**Review by bboreham — COMMENTED** [review]
I can't see where this PR does anything worthwhile.
Postings returns an iterator - not much is done in the initialization alone.
Maybe I'm missing something?

File: tsdb/index/postings.go (hunk 1 above)
Comment (aknuds1) [reply tsdb/index/postings.go:570]: I think I agree in retrospect, that you don't gain much here. My thinking was probably that `it.Next()` is potentially slow?

**Pull request merged**